shared/src/commonMain/kotlin/org/hyperskill/app/track_selection/details/presentation/TrackSelectionDetailsReducer.kt

```diff
@@ -0,0 +1,114 @@
+    override fun reduce(state: State, message: Message): Pair<State, Set<Action>> =
+        when (message) {
+            is Message.Initialize -> {
+                state.updateContentState(ContentState.Loading) to
+                    fetchAdditionalInfo(state, forceLoadFromNetwork = false)
+            }
+            is Message.RetryContentLoading ->
+                handleRetryContentLoading(state)
+            is TrackSelectionDetailsFeature.FetchProvidersAndFreemiumStatusResult ->
+                handleAdditionalInfoFetchResult(state, message)
+            Message.SelectTrackButtonClicked ->
+                handleSelectTrackButtonClicked(state)
+            is TrackSelectionDetailsFeature.TrackSelectionResult ->
+                handleTrackSelectionResult(state, message)
+            is Message.ViewedEventMessage ->
+                state to setOf(
+                    InternalAction.LogAnalyticEvent(
+                        TrackSelectionDetailsViewedHyperskillAnalyticEvent(state.trackWithProgress.track.id)
+                    )
+                )
+        }
+
+    private fun handleRetryContentLoading(
+        state: State
+    ): ReducerResult {
+        val logAnalyticsAction = InternalAction.LogAnalyticEvent(
+            TrackSelectionDetailsClickedRetryContentLoadingHyperskillAnalyticsEvent(
+                trackId = state.trackWithProgress.track.id
+            )
+        )
+        return if (state.contentState is ContentState.NetworkError) {
+            state.updateContentState(ContentState.Loading) to
+                fetchAdditionalInfo(state, forceLoadFromNetwork = true) + logAnalyticsAction
+        } else {
+            state to setOf(logAnalyticsAction)
+        }
+    }
+
+    private fun fetchAdditionalInfo(
+        state: State,
+        forceLoadFromNetwork: Boolean
+    ): Set<Action> =
+        setOf(
+            InternalAction.FetchProvidersAndFreemiumStatus(
+                providerIds = state.trackWithProgress.track.topicProviders,
+                forceLoadFromNetwork = forceLoadFromNetwork
+            )
+        )
+
+    private fun handleAdditionalInfoFetchResult(
+        state: State,
+        message: TrackSelectionDetailsFeature.FetchProvidersAndFreemiumStatusResult
+    ): ReducerResult {
+        val contentState = when (message) {
+            is TrackSelectionDetailsFeature.FetchProvidersAndFreemiumStatusResult.Success -> {
+                ContentState.Content(
+                    isFreemiumEnabled = message.isFreemiumEnabled,
+                    providers = message.providers
+                )
+            }
+            TrackSelectionDetailsFeature.FetchProvidersAndFreemiumStatusResult.Error ->
+                ContentState.NetworkError
+        }
+        return state.updateContentState(contentState) to emptySet()
+    }
+
+    private fun handleSelectTrackButtonClicked(
+        state: State
+    ): ReducerResult =
+        state.copy(isTrackLoadingShowed = true) to
```

Comment:
Here you should check for `isTrackSelected` and if `true` block the ability to select an already selected track in case of platform UI is not blocked `Select this track` button

Reply:
@ivan-magda 
I've also added the explicit `isSelectTrackButtonEnabled` property to the ViewState.